runtime/gc_base/GCExtensions.cpp

```diff
@@ -267,3 +273,71 @@ MM_GCExtensions::computeDefaultMaxHeap(MM_EnvironmentBase *env)
 
 	memoryMax = MM_Math::roundToFloor(heapAlignment, memoryMax);
 }
+
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+#if !defined(LINUX) || !defined(J9VM_ENV_DATA64) || defined(OMRZTPF)
+/* Double map is only supported on LINUX 64 bit Systems for now */
+#error "Platform not supported by Double Map API"
+#endif /* !LINUX || !defined(J9VM_ENV_DATA64) || defined(OMRZTPF) */
+void* 
+MM_GCExtensions::doubleMapArraylets(MM_EnvironmentBase* env, J9Object *objectPtr) 
+{
+	J9JavaVM *javaVM = getJavaVM();
+	PORT_ACCESS_FROM_ENVIRONMENT(env);
+
+	GC_ArrayletLeafIterator arrayletLeafIterator(javaVM, (J9IndexableObject*)objectPtr);
+	UDATA arrayletLeafCount = arrayletLeafIterator.getNumLeafs();
+	MM_Heap *heap = getHeap();
+
+	void *result = NULL;
+	void **arrayletLeaveAddrs = (void **)j9mem_allocate_memory(arrayletLeafCount * sizeof(uintptr_t), J9MEM_CATEGORY_VM);
+	if(arrayletLeaveAddrs == NULL) {
+		return NULL;
+	}
+	UDATA elementsSize = indexableObjectModel.getDataSizeInBytes((J9IndexableObject*)objectPtr);
+
+	GC_SlotObject *slotObject = NULL;
+	uintptr_t count = 0;
+
+	while (NULL != (slotObject = arrayletLeafIterator.nextLeafPointer())) {
+		void *currentLeaf = slotObject->readReferenceFromSlot();
+		arrayletLeaveAddrs[count] = currentLeaf;
+		count++;
+	}
+
+	GC_ArrayletLeafIterator arrayletLeafIterator2(javaVM, (J9IndexableObject*)objectPtr);
+    GC_SlotObject *slotObj = arrayletLeafIterator2.nextLeafPointer();
+    MM_HeapRegionDescriptorVLHGC *dest = (MM_HeapRegionDescriptorVLHGC *)heap->getHeapRegionManager()->tableDescriptorForAddress(slotObj->readReferenceFromSlot());
+
+	UDATA arrayletLeafSize = javaVM->arrayletLeafSize;
+	UDATA pageSize = j9mmap_get_region_granularity(NULL); /* gets pagesize  or j9vmem_supported_page_sizes()[0]? */
+
+	// Get heap and from there call an OMR API that will doble map everything
+	result = heap->doubleMapArraylet(env, arrayletLeaveAddrs, count, arrayletLeafSize, elementsSize, 
+				&dest->_identifier,
+				pageSize);
+
+	if (NULL == result) { /* Double map failed */
+		return NULL;
+	}
```

Comment:
not part of this change, but need to gather some stats on how many discontiguous arrays we allocated and than how many (hopefully never) failed to double map.
we already have allocate-stats stanza in Verbose GC (that reports some basic alloc stats between) two GC cycles. this is where it would be reported.
currently, we do report number of leafs that are live (in mem-info stanza), but not alloc stats. both are interesting.